### PR TITLE
Feat/exercises v2

### PR DESCRIPTION
**Pull request by MuskanThakur2002**

### CODE REVIEW

File: lib/dbTriggers/index.js [hunk 1]
```diff
@@ -168,8 +168,8 @@ async function checkPathwayCompletionEligibility(transaction, userId, id, versio
       promisesAllCourses.push(allCourses);
       const completedCourses = CourseCompletionV2.query(transaction)
         .where('user_id', userId)
-        .withGraphJoined('pathwayCourses')
-        .where('pathwayCourses.pathway_id', pathwayIdArr[i]);
+        .withGraphJoined('pathwayCoursesV2')
```
Comment (pysaquib) [inline lib/dbTriggers/index.js:171]: Keep this

File: lib/courseParser/index.js [hunk 2]
```diff
@@ -1721,7 +1721,7 @@ class CoursesSeeder {
               });
             } else {
               const listItemstype = listItems.raw.trim().split('')[0];
-              listText = listText.replace(/\n/g, '<br>');
+              listText = listText.replace(/\n/g, '');
```
Comment (pysaquib) [inline lib/courseParser/index.js:1724]: Confirm this once with @AnandDhakane01 . He made this change.

File: lib/courseParser/index.js [hunk 3]
```diff
@@ -1686,7 +1686,7 @@ class CoursesSeeder {
                       });
                     } else {
                       const innerListItemtype = listAtomicToken.raw.trim().split('')[0];
-                      listAtomicToken.text = listAtomicToken.text.replace(/\n/g, '<br>');
+                      listAtomicToken.text = listAtomicToken.text.replace(/\n/g, '');
```
Comment (pysaquib) [inline lib/courseParser/index.js:1689]: Confirm wit @AnandDhakane01

File: lib/courseParser/index.js [hunk 4]
```diff
@@ -1384,7 +1384,7 @@ class CoursesSeeder {
           // if list items are MD links
           if (listItems.text.match(/.*(\[.+\])(\((http).+\)).*/g)) {
             const listType = listItems.raw.trim().split(' ')[0];
-            listItems.text = listItems.text.replace(/\n/g, '<br>');
+            listItems.text = listItems.text.replace(/\n/g, '');
```
Comment (pysaquib) [inline lib/courseParser/index.js:1387]: Confirm with @AnandDhakane01

File: lib/_mockAPIs/classTopic.json [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+            }
+          ],
+          "content_type": "assessment"
+        }
+      ],
+      "description": "Let's learn typing"
+    }
+  }
```
Comment (pysaquib) [inline lib/_mockAPIs/classTopic.json:354]: remove this file

File: lib/dbTriggers/index.js [hunk 6]
```diff
@@ -168,8 +168,8 @@ async function checkPathwayCompletionEligibility(transaction, userId, id, versio
...
+        .where('pathwayCoursesV2.pathway_id', pathwayIdArr[i]);
```
Comment (pysaquib) [inline lib/dbTriggers/index.js:172]: Keep these lines

File: lib/migrations/20211116132741_courses_v2.js [hunk 7]
```diff
@@ -10,7 +10,7 @@ exports.up = async (knex) => {
     table.increments().primary();
     table.string('name').notNullable();
     table.string('logo');
-    table.string('description');
+    table.string('short_description');
```
Comment (pysaquib) [inline lib/migrations/20211116132741_courses_v2.js:13]: Keep this. Let's keep `short_description` in both `courses` and `courses_v2`

File: lib/migrations/20211116132741_courses_v2.js [hunk 8]
```diff
@@ -41,12 +41,19 @@ exports.up = async (knex) => {
       .unsigned()
       .references('id')
       .inTable('main.pathways_v2')
-      .notNullable();
+      .notNullable()
+      .unique();
```
Comment (pysaquib) [inline lib/migrations/20211116132741_courses_v2.js:45]: Keep this

File: lib/migrations/20211116132741_courses_v2.js [hunk 9]
```diff
@@ -41,12 +41,19 @@ exports.up = async (knex) => {
...
     table.increments().primary();
     table.integer('user_id').unsigned().references('id').inTable('main.users').notNullable();
-    table.integer('course_id').unsigned().references('id').inTable('main.courses_v2').notNullable();
+    table
+      .integer('course_id')
+      .unsigned()
+      .references('id')
+      .inTable('main.courses_v2')
+      .notNullable()
+      .unique();
```
Comment (pysaquib) [inline lib/migrations/20211116132741_courses_v2.js:56]: Keep this

File: lib/migrations/20211116132741_courses_v2.js [hunk 10]
```diff
@@ -56,7 +63,8 @@ exports.up = async (knex) => {
       .unsigned()
       .references('id')
       .inTable('main.exercises_v2')
-      .notNullable();
+      .notNullable()
+      .unique();
```
Comment (pysaquib) [inline lib/migrations/20211116132741_courses_v2.js:67]: Keep this

File: lib/models/coursesV2.js [hunk 11]
```diff
@@ -12,7 +12,7 @@ module.exports = class CoursesV2 extends ModelBase {
       id: Joi.number().integer().greater(0),
       name: Joi.string().required(),
       logo: Joi.string().required(),
-      description: Joi.string().required(),
+      short_description: Joi.string().required(),
```
Comment (pysaquib) [inline lib/models/coursesV2.js:15]: Keep this

File: lib/models/user.js [hunk 12]
```diff
@@ -36,6 +36,7 @@ module.exports = class User extends ModelBase {
     const Classes = require('./classes');
     const ClassRegistrations = require('./classRegistrations');
     const StudentPathways = require('./studentPathway');
+    const Partner = require('./partner');
```
Comment (pysaquib) [inline lib/models/user.js:39]: Keep this

File: lib/models/user.js [hunk 13]
```diff
@@ -97,6 +98,14 @@ module.exports = class User extends ModelBase {
...
         },
       },
+      partner: {
+        relation: Model.HasManyRelation,
+        modelClass: Partner,
+        join: {
+          from: 'main.users.partner_id',
+          to: 'main.partners.id',
+        },
+      },
```
Comment (pysaquib) [inline lib/models/user.js:108]: Keep this

File: lib/routes/courses.js [hunk 14]
```diff
@@ -1,7 +1,12 @@
 const Joi = require('@hapi/joi');
+const fs = require('fs');
+const path = require('path');
 const Courses = require('../models/courses');
 const logger = require('../../server/logger');
 
+const androidVersions = {
+  latest: 26,
+};
```
Comment (pysaquib) [inline lib/routes/courses.js:9]: Keep this

File: lib/routes/courses.js [hunk 15]
```diff
@@ -69,6 +115,21 @@ module.exports = [
...
+        if (courseId === 1000) {
+          const data = fs.readFileSync(path.resolve(__dirname, '../_mockAPIs/classTopic.json'));
+          return data.toString();
+        }
+        /**
+         * EXPERIMENTAL
+         *
+         * TO BE REMOVED
+         */
+
```
Comment (pysaquib) [inline lib/routes/courses.js:132]: Remove this entire part from line 118 - 132

**Review by pysaquib — COMMENTED** [review]
Checked

File: lib/courseParser/index.js (hunk 2 above)
Comment (MuskanThakur2002) [reply lib/courseParser/index.js:1724]: okay

**Pull request merged**